jinahub/rankers/LightGBMRanker/lightgbm_ranker.py

```diff
@@ -0,0 +1,209 @@
+    .. note::
+        The name of the features are used to extract the features from incoming `documents`. Check how these features are accessed in
+        :class:`Document` at https://docs.jina.ai/api/jina.types.document/
+    """
+
+    def __init__(
+        self,
+        query_features: List[str],
+        match_features: List[str],
+        relevance_label: str,
+        model_path: Optional[str] = None,
+        params: Dict = {
+            'task': 'train',
+            'boosting_type': 'gbdt',
+            'objective': 'lambdarank',
+            'min_data_in_leaf': 1,
+            'feature_pre_filter': False,
+        },
+        categorical_query_features: Optional[List[str]] = None,
+        categorical_match_features: Optional[List[str]] = None,
+        query_features_before: bool = False,
+        *args,
+        **kwargs,
+    ):
+        super(LightGBMRanker, self).__init__(*args, **kwargs)
+        self.params = params
+        self.model_path = model_path
+        self.query_features = query_features
+        self.match_features = match_features
+        self.categorical_query_features = categorical_query_features
+        self.categorical_match_features = categorical_match_features
+        self.query_features_before = query_features_before
+        self.label = relevance_label
+        self.logger = JinaLogger(self.__class__.__name__)
+        if self.model_path and os.path.exists(self.model_path):
+            self.booster = self._load_model(self.model_path)
+        else:
+            self.booster = None
+
+    def _load_model(self, path):
+        """Load model from model path"""
+        self.booster = lightgbm.Booster(model_file=path)
+        model_num_features = self.booster.num_feature()
+        expected_num_features = len(self.query_features + self.match_features)
+        if self.categorical_query_features:
+            expected_num_features += len(self.categorical_query_features)
+        if self.categorical_match_features:
+            expected_num_features += len(self.categorical_match_features)
+        if model_num_features != expected_num_features:
+            raise ValueError(
+                f'The number of features expected by the LightGBM model {model_num_features} is different'
+                f'than the ones provided in input {expected_num_features}'
+            )
+
+    def _save_model(self, path):
+        """Dump model into cbm."""
+        self.booster.save_model(path)
+
+    def _get_features_dataset(self, docs: DocumentArray) -> 'lightgbm.Dataset':
+        q_features, m_features, group, labels = [], [], [], []
+        query_feature_names = self.query_features
+        if self.categorical_query_features:
+            query_feature_names += self.categorical_query_features
+        match_feature_names = self.match_features
+        if self.categorical_match_features:
+            match_feature_names += self.categorical_match_features
+        for doc in docs:
```

Comment:
I would refactor this - use `docs.get_attributes(f'tags__{tag_name}')`, instead of looping over documents

Reply:
Nevermind actually - I see you are doing it because of the matches